include/dlaf/matrix/tile.h

```diff
@@ -338,36 +379,77 @@ class Tile : public Tile<const T, D> {
     return data_.ptr(index);
   }
 
-  /// Sets the promise to which this Tile will be moved on destruction.
-  ///
-  /// @c setPromise can be called only once per object.
-  /// @pre this Tile should not be a subtile.
-  Tile& setPromise(TilePromise&& p) {
-    DLAF_ASSERT(!std::holds_alternative<TilePromise>(dep_tracker_),
-                "setPromise has been already used on this object!");
-    DLAF_ASSERT(std::holds_alternative<std::monostate>(dep_tracker_),
-                "setPromise cannot be used on subtiles!");
-    dep_tracker_ = std::move(p);
-    return *this;
-  }
-
   /// Returns a subtile.
   /// Note: to avoid segfaults or race conditions, the original tile must be kept in scope.
   Tile subTileReference(const SubTileSpec& spec) const noexcept {
     return Tile(*this, spec);
   }
 
 private:
-  // Creates an untracked subtile.
-  // Dependencies are not influenced by the new created object therefore race-conditions
-  // might happen if used improperly.
   Tile(const Tile& tile, const SubTileSpec& spec) noexcept : Tile<const T, D>(tile, spec) {}
 
-  // Creates a writable subtile keeping the dependencies.
-  // It calls old_tile.get(), therefore it should be used when old_tile is guaranteed to be ready:
-  // e.g. in dataflow, .then, ...
-  Tile(pika::shared_future<TileType> tile, const SubTileSpec& spec) : ConstTileType(tile.get(), spec) {
-    dep_tracker_ = std::move(tile);
+  Tile(internal::TileAsyncRwMutexReadOnlyWrapper<T, D> tile_wrapper, const SubTileSpec& spec)
+      : Tile<const T, D>(tile_wrapper.get(), spec) {
+    dep_tracker_ = std::move(tile_wrapper);
+  }
+
+  Tile(internal::TileAsyncRwMutexReadWriteWrapper<T, D> tile_wrapper)
+      : ConstTileType(tile_wrapper.get().size(), tile_wrapper.get().data_.memoryView(),
+                      tile_wrapper.get().ld()) {
+    dep_tracker_ = std::move(tile_wrapper);
+  }
+
+  Tile(internal::TileAsyncRwMutexReadWriteWrapper<T, D> tile_wrapper, const SubTileSpec& spec)
+      : ConstTileType(tile_wrapper.get(), spec) {
+    dep_tracker_ = std::move(tile_wrapper);
+  }
+
+  Tile(Tile&& tile, const SubTileSpec& spec) : ConstTileType(tile, spec) {
+    dep_tracker_ = std::move(tile.dep_tracker_);
+  }
+
+  void prepareDisjointTile() {
+    // We only expect read-write tiles for disjoint access. That means that the
+    // dependency tracker holds anything but a read-only wrapper.
+    // TODO: Condition is not inside DLAF_ASSERT because it contains a comma. Do
+    // we want to support a more convenient syntax for this or is this ok?
```

Comment:
@rasolca, @albestro are you fine with me removing this TODO? Or do you have suggestions for how to improve this?

Reply:
I'm ok with that...I don't remember if just wrapping inside an additional pair of `()` worked or not in this kind of situations.

Reply:
Hmm, good point. Let me try that quickly.

Reply:
Brilliant @albestro, thanks for pointing that out. See https://github.com/eth-cscs/DLA-Future/pull/766/commits/dd9dbe79322ab5c0810a163574dd01f92c766f2a.